test/e2e/catalog_e2e_test.go

```diff
@@ -1056,27 +1057,31 @@ var _ = Describe("Catalog represents a store of bundles which OLM can use to ins
 		}
 
 		By("creating a catalog source")
-		source, err := crc.OperatorsV1alpha1().CatalogSources(source.GetNamespace()).Create(context.TODO(), source, metav1.CreateOptions{})
-		Expect(err).ToNot(HaveOccurred())
-
-		By("update the catalog source with template annotation")
 
-		source, err = crc.OperatorsV1alpha1().CatalogSources(source.GetNamespace()).Get(context.TODO(), source.GetName(), metav1.GetOptions{})
-		Expect(err).ShouldNot(HaveOccurred(), "error getting catalog source")
-
-		// create an annotation using the kube templates
+		var err error
+		Eventually(func() error {
+			source, err = crc.OperatorsV1alpha1().CatalogSources(source.GetNamespace()).Create(context.TODO(), source, metav1.CreateOptions{})
+			return err
+		}).Should(Succeed())
 
-		source.SetAnnotations(map[string]string{catalogsource.CatalogImageTemplateAnnotation: fmt.Sprintf("quay.io/olmtest/catsrc-update-test:%s.%s.%s", catalogsource.TemplKubeMajorV, catalogsource.TemplKubeMinorV, catalogsource.TemplKubePatchV)})
+		By("update the catalog source with template annotation")
 
 		// Update the catalog image
 		Eventually(func() error {
+			source, err = crc.OperatorsV1alpha1().CatalogSources(source.GetNamespace()).Get(context.TODO(), source.GetName(), metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			// create an annotation using the kube templates
+			source.SetAnnotations(map[string]string{catalogsource.CatalogImageTemplateAnnotation: fmt.Sprintf("quay.io/olmtest/catsrc-update-test:%s.%s.%s", catalogsource.TemplKubeMajorV, catalogsource.TemplKubeMinorV, catalogsource.TemplKubePatchV)})
+
```

Comment:
Revisiting this as it's been sitting for a while: poking at this locally, it looks like the real issue is we're overwriting this shared `source` variable in the GET and UPDATE calls. Speculatively, it looks like anytime the update call fails, the source variable will be essentially nil:

```
catalogsource info: catalog-9t8pz/openshift-operators
catalogsource info: &v1alpha1.CatalogSource{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"catalog-9t8pz", GenerateName:"", Namespace:"openshift-operators", SelfLink:"", UID:"1151883a-39f2-44a0-9d42-8c2e28505b47", ResourceVersion:"32421", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63772780625, loc:(*time.Location)(0x46b9aa0)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"olm.catalogSource":"catalog-9t8pz"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"e2e.test", Operation:"Update", APIVersion:"operators.coreos.com/v1alpha1", Time:(*v1.Time)(0xc0001be498), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0xc0001be480), Subresource:""}}}, Spec:v1alpha1.CatalogSourceSpec{SourceType:"grpc", Priority:0, ConfigMap:"", Address:"", Image:"quay.io/olmtest/catsrc-update-test:old", UpdateStrategy:(*v1alpha1.UpdateStrategy)(nil), Secrets:[]string(nil), DisplayName:"", Description:"", Publisher:"", Icon:v1alpha1.Icon{Data:"", MediaType:""}}, Status:v1alpha1.CatalogSourceStatus{Message:"", Reason:"", LatestImageRegistryPoll:(*v1.Time)(nil), ConfigMapResource:(*v1alpha1.ConfigMapResourceReference)(nil), RegistryServiceStatus:(*v1alpha1.RegistryServiceStatus)(nil), GRPCConnectionState:(*v1alpha1.GRPCConnectionState)(nil), Conditions:[]v1.Condition(nil)}}
catalogsource info: /
catalogsource info: /
...
```

One solution is to just avoid overwriting this source resource by suppressing the variable returned from the update call. I tested out this variation locally and I wasn't able to reproduce this issue anymore.